src/plugins/visualizations/public/embeddable/to_ast.ts

```diff
@@ -22,15 +23,24 @@ import { queryToAst, filtersToAst } from '../../../data/common';
 export const toExpressionAst: VisToExpressionAst = async (vis, params) => {
   const { savedSearchId, searchSource } = vis.data;
   const query = searchSource?.getField('query');
-  let filters = searchSource?.getField('filter');
+  let filters = searchSource?.getField('filter') || [];
   if (typeof filters === 'function') {
-    filters = filters();
+    filters = filters() || [];
+  }
+  if (filters && !Array.isArray(filters)) {
+    filters = [filters];
   }
 
   const kibana = buildExpressionFunction<ExpressionFunctionKibana>('kibana', {});
   const kibanaContext = buildExpressionFunction<ExpressionFunctionKibanaContext>('kibana_context', {
     q: query && queryToAst(query),
-    filters: filters && filtersToAst(filters),
+    filters:
+      filters &&
+      filtersToAst(
+        filters.map((filter) => {
+          return migrateFilter(filter);
```

Comment:
Do we need to do this for Lens as well?

Reply:
you should be OK if:
if you store filters in SO and correctly use persistabestate migrate utilities 
if you store filters in URL together with kibanaVersion and correctly use persistablestate migrate utilities

its also OK if (but you should not rely on this):
if you store filters in URL and populate filterManager with them before using them
if you store filters in URL and use them with our high level search service or expressions
if you store filters in URL and use convert them using our utilities before passing them to low level search service.